src/parsers.rs

```diff
@@ -115,6 +136,110 @@ impl<'a> Parser<'a> {
         Err(())
     }
 
+    #[cfg(feature = "ipv6")]
+    fn accept_ipv6_part(&mut self,
+      (head, tail): (&mut [u16; 8], &mut [u16; 6]),
+      (head_idx, tail_idx): (&mut usize, &mut usize),
+      mut use_tail: bool) -> Result<()> {
+        let mut double_colon = false;
+        match self.try(|p| p.accept_str(b"::")) {
+            Some(_) if !use_tail && *head_idx < 7 => {
+                // Found a double colon. Start filling out the
+                // tail and set the double colon flag in case
+                // this is the last character we can parse.
+                use_tail = true;
+                double_colon = true;
+            },
+            Some(_) => {
+                // This is a bad address. Only one double colon is
+                // allowed and an address is only 128 bits.
+                return Err(());
+            }
+            None if *head_idx != 0 || use_tail && *tail_idx != 0 => {
+                // If this is not the first number or the position following
+                // a double colon, we expect there to be a single colon.
+                self.accept_char(b':')?;
+            }
+            None => (),
+        };
+
+        match self.try(|p| p.accept_number(4, 0x10000, true)) {
+            Some(part) if !use_tail && *head_idx < 8 => {
+                // Valid u16 to be added to the address
+                head[*head_idx] = part as u16;
+                *head_idx += 1;
+                Ok(())
+            },
+            Some(part) if *tail_idx < 6 => {
+                // Valid u16 to be added to the address
+                tail[*tail_idx] = part as u16;
+                *tail_idx += 1;
+                Ok(())
+            },
+            Some(_) => {
+                // Tail or head section is too long
+                Err(())
+            }
+            None if self.lookahead_char(b'/') => {
```

Comment:
No, this is not right. You shouldn't look for any specific character. Rather, you should bail out if you detect anything you don't expect.

Reply:
This particular branch is really only applicable to CIDRs, so maybe I should reintroduce `is_cidr` then